### ISSUE TITLE
Minio gateway with encryption doesn't work auto-encryption enabled

### ISSUE THREAD

**Issue by annkam** [minio/minio#12335]
Good afternoon! We've set up a minio gateway with encryption in front of a basic Minio server with no encryption. We want all our data to be encrypted only at the gateway, so in case of the direct access to the backend we can not view the data unencrypted

This configuration works:

```
version: '2.4' 
  
services: 
 miniogw: 
   image: minio/minio:RELEASE.2021-04-22T15-44-28Z 
   container_name: minio-gw 
   environment: 
     MINIO_ROOT_USER: miniogw 
     MINIO_ROOT_PASSWORD: minio123 
     AWS_ACCESS_KEY_ID: minio 
     AWS_SECRET_ACCESS_KEY: minio123 
     MINIO_KMS_KES_ENDPOINT: https://play.min.io:7373 
     MINIO_KMS_KES_CERT_FILE: /minio-data/root.cert 
     MINIO_KMS_KES_KEY_FILE: /minio-data/root.key 
     MINIO_KMS_KES_KEY_NAME: my-app-key 
     MINIO_KMS_AUTO_ENCRYPTION: "on" 
     MINIO_GATEWAY_SSE: "c" 
   volumes: 
     - ./data:/minio-data 
   ports: 
     - "9001:9000" 
   command: gateway s3 http://minio:9000 
   depends_on: 
     minio: 
       condition: service_healthy 
 minio: 
   image: minio/minio:RELEASE.2021-05-18T00-53-28Z
   container_name: minio 
   environment: 
     MINIO_ROOT_USER: minio 
     MINIO_ROOT_PASSWORD: minio123 
   command: server /minio-data/data 
   volumes: 
     - ./data:/minio-data 
   ports: 
     - "9000:9000" 
   healthcheck: 
     test: "curl -sS http://localhost:9000/minio/health/live || echo 1" 
     interval: 5s 
     timeout: 10s 
     retries: 3
```

Notice that the gateway is on version `RELEASE.2021-04-22T15-44-28Z` and the server is on `RELEASE.2021-05-18T00-53-28Z`. After upgrading to `RELEASE.2021-05-18T00-53-28Z` the gateway does not work.

## Expected Behavior

I expect gateway encryption to work on latest release

## Current Behavior

Objects can't be uploaded with the following error:

```
minio-gw   | API: SYSTEM() 
minio-gw   | Time: 08:09:06 UTC 05/20/2021 
minio-gw   | DeploymentID: bfca7554-018d-4fe7-a7c1-d03788911aba 
minio-gw   | Error: Server side encryption specified but KMS is not configured (minio.ErrorResponse) 
minio-gw   |        5: cmd/api-errors.go:2031:cmd.toAPIErrorCode() 
minio-gw   |        4: cmd/api-errors.go:2056:cmd.toAPIError() 
minio-gw   |        3: cmd/web-handlers.go:2446:cmd.toWebAPIError() 
minio-gw   |        2: cmd/web-handlers.go:2455:cmd.writeWebErrorResponse() 
minio-gw   |        1: cmd/web-handlers.go:1327:cmd.(*webAPIHandlers).Upload()
```

## Steps to Reproduce
1) Deploy provided docker-compose file
2) Test uploading/downloading files
3) Change miniogw image tag to RELEASE.2021-05-18T00-53-28Z
4) Try to upload a file

**Comment by klauspost** [comment]
> `MINIO_KMS_KES_ENDPOINT: https://play.min.io:7373`

It appears using the `https://play.min.io` (our server) as a key server. I presume you have set up your own? 

Ping @aead

**Comment by annkam** [comment]
Yes, we used play.min.io as an example just to make sure that the problem is not with our KES server

**Comment by vadmeste** [comment]
@annkam are you saying that migrating the gateway RELEASE.2021-04-22T15-44-28Z  to RELEASE.2021-05-18T00-53-28Z broke the gateway encryption ?

**Comment by aead** [comment]
This may be a side-effect of switching the default SSE method from SSE-S3 to SSE-KMS. Currently the gateway only supports SSE-S3 and SSE-C but `MINIO_KMS_AUTO_ENCRYPTION=on` now (automatically) adds SSE-KMS headers if no SSE headers are present.

@annkam Can you confirm that adding the SSE-S3 header (`X-Amz-Server-Side-Encryption: AES256`) at the client explicitly works but without that header it fails?

**Comment by aead** [comment]
In this case we would need to add SSE-KMS support to the gateway here: https://github.com/minio/minio/blob/master/cmd/gateway-env.go#L27

**Comment by harshavardhana** [comment]
@aead let's discuss if we need to provide support for gateway. If it's already removed we can move forward and do not have to support this use case anymore. Perhaps even purge the entire SSE layer for gateway.

You are free to not upgrade your gateway instance @annkam we are removing gateway from server and moving it to https://github.com/minio/ming anyways.

**Comment by annkam** [comment]
@vadmeste Yes, upgrade breaks it

@aead I will check adding the header

@harshavardhana What do you mean by "purge the entire SSE layer for gateway"? Are you going to remove the entire support of encryption for the gateway?  Will the new gateway support ETCD?

**Comment by harshavardhana** [comment]
> @harshavardhana What do you mean by "purge the entire SSE layer for gateway"? Are you going to remove the entire support of encryption for the gateway?  Will the new gateway support ETCD? 
> 

The entire layer that exists in this codebase will be removed. We will bring the support back into ming in a slightly different manner.

The ming will support what we currently support in minio gateway 

This removal comes with the overall simplification for the server code base.

**Comment by annkam** [comment]
@aead  I confirm that it works with header `X-Amz-Server-Side-Encryption: AES256`
What about `KMS_AUTO_ENCRYPTION` env var? Is it okay to use it for auto-encryption on the Gateway? Will it (or a similar var) be present in ming project?

**Comment by harshavardhana** [comment]
> @aead I confirm that it works with header `X-Amz-Server-Side-Encryption: AES256`
> What about `KMS_AUTO_ENCRYPTION` env var? Is it okay to use it for auto-encryption on the Gateway? Will it (or a similar var) be present in ming project?

The answer to that 

> The ming will support what we currently support in minio gateway

**Comment by harshavardhana** [comment]
Should be fixed in #13257
